packages/react/src/hooks/useAnchoredPosition.ts

```diff
@@ -30,22 +31,49 @@ export function useAnchoredPosition(
   const floatingElementRef = useProvidedRefOrCreate(settings?.floatingElementRef)
   const anchorElementRef = useProvidedRefOrCreate(settings?.anchorElementRef)
   const [position, setPosition] = React.useState<AnchorPosition | undefined>(undefined)
+  const [_, setPrevHeight] = React.useState<number | undefined>(undefined)
 
   const updatePosition = React.useCallback(
     () => {
       if (floatingElementRef.current instanceof Element && anchorElementRef.current instanceof Element) {
-        setPosition(getAnchoredPosition(floatingElementRef.current, anchorElementRef.current, settings))
+        const newPosition = getAnchoredPosition(floatingElementRef.current, anchorElementRef.current, settings)
+        setPosition(prev => {
+          if (
+            settings?.pinPosition &&
+            prev &&
+            ['outside-top', 'inside-top'].includes(prev.anchorSide) &&
+            (prev.anchorSide !== newPosition.anchorSide || prev.top < newPosition.top)
+          ) {
+            const anchorTop = anchorElementRef.current?.getBoundingClientRect().top ?? 0
+            if (anchorTop > (floatingElementRef.current?.clientHeight ?? 0)) {
+              setPrevHeight(prevHeight => {
+                if (prevHeight && prevHeight > (floatingElementRef.current?.clientHeight ?? 0)) {
+                  requestAnimationFrame(() => {
+                    ;(floatingElementRef.current as HTMLElement).style.height = `${prevHeight}px`
+                  })
+                } else {
+                  prev = newPosition
```

Comment:
edgecase where the selectPanel is actually growing instead of shrinking, in this case we do want to allow it to recalculate/reposition